lib/core/notifications/components/notification-history.component.ts

```diff
@@ -47,43 +44,100 @@ export class NotificationHistoryComponent implements OnDestroy {
     @Input()
     menuPositionY: string = 'below';
 
+    onDestroy$ = new Subject<boolean>();
+    notifications: NotificationModel[] = [];
+    paginatedNotifications = [];
+    pagination: Pagination;
+
     constructor(
-        private notificationService: NotificationService, public storageService: StorageService) {
-        this.notifications = JSON.parse(storageService.getItem('notifications')) || [];
+        private notificationService: NotificationService,
+        public storageService: StorageService,
+        public cd: ChangeDetectorRef) {
+
+    }
+
+    ngOnInit() {
+        this.notifications = JSON.parse(this.storageService.getItem('notifications')) || [];
+    }
+
+    ngAfterViewInit(): void {
         this.notificationService.notifications$
-            .pipe(takeUntil(this.onDestroy$))
-            .subscribe((message) => {
-                this.notifications.push(message);
+        .pipe(takeUntil(this.onDestroy$))
+        .subscribe((notification: NotificationModel) => {
+            this.addNewNotification(notification);
+            this.cd.detectChanges();
+        });
+    }
+
+    ngOnDestroy() {
+        this.onDestroy$.next(true);
+        this.onDestroy$.complete();
+    }
 
-                if (this.notifications.length > this.MAX_NOTIFICATION_STACK_LENGTH) {
-                    this.notifications.shift();
-                }
+    addNewNotification(notification: NotificationModel) {
+        this.notifications.unshift(notification);
+
+        if (this.notifications.length > NotificationHistoryComponent.MAX_NOTIFICATION_STACK_LENGTH) {
+            this.notifications.shift();
+        }
 
-                storageService.setItem('notifications', JSON.stringify(this.notifications));
-            });
+        this.saveNotifications();
+        this.resetPagination();
+        this.updatePagination();
     }
 
-    isEmptyNotification(): boolean {
-        return (!this.notifications || this.notifications.length === 0);
+    saveNotifications() {
+        this.storageService.setItem('notifications', JSON.stringify(this.notifications.filter((notification) =>
+            notification.type !== NOTIFICATION_TYPE.RECURSIVE
+        )));
+    }
+
+    onMenuOpened() {
+        this.resetPagination();
```

Comment:
why do we need to reset and update?

Reply:
to reset the pagination, otherwise if you open the menu, go through 100 notifications loading more and more, next time you open it you'll get the 100 notifications right away. This way, when a user clicks on the menu, he'll only see a maximum on 5 tasks before loading more